source/core/slang-tree-map.h

```diff
@@ -0,0 +1,474 @@
+#ifndef SLANG_CORE_TREE_MAP_H
+#define SLANG_CORE_TREE_MAP_H
+
+#include "slang-list.h"
+#include "slang-common.h"
+#include <initializer_list>
+
+namespace Slang
+{
+
+    // Array-backed Red-black Tree Map.
+    template<typename T>
+    class TreeMap
```

Comment:
> My guess is that you want to have the result to be in a sorted order

This is correct. I want the result to be in sorted over after every "add" operation.

> In practice, it is more efficient and simple to collect elements in a random order and do one time sort at the end; if there aren't duplicates.

Sorting a `List` after adding a new element will likely be slower in our use-case. This is because the code using `TreeSet` does the following 3 functionalities:
1. Distinct pass over elements where we 'search for an element', then we 'add a new element',  `TreeSet` is significantly faster to search than an unordered list, adding elements is slower.
2. [If we were using a `List` we would sort here]
3. We in a loop then search from least to greatest and insert a sorted element, `List` will likely be slower to insert a sorted element, `List` will be slightly faster to iterate.

> If we will have this container, the naming should change to Set, because Map implies a mapping from A to B.

I agree.